Fix tests/lib-threads/beat.ml on macOS ARM

dra27 opened the pull request:
This test was reported during the development of multicore as being "off-by-one", which was assumed to be an issue with multicore at the time (see ocaml-multicore/ocaml-multicore#408). However, as macOS CI runners have become more generally available, it's become more apparent in OCaml 4.x, especially on macOS ARM hardware in CI, where the test consistently fails with the first parameter often coming in at 7 (outside the acceptable range of 8-10).

The underlying issue is not OCaml, but `nanosleep`, which appears to be capable of adding up to an additional 30% to the given wait time for sleep intervals less than 1 second. This is occasionally seen on the amd64 macOS runners as well.

The previous computation assumed that the first thread (with a tick of 0.333s) should delay 9±1 times during the 3s runtime and the second thread (with a tick of 0.5s) should delay 6±1 times.

To account for the extra delay at least on macOS, we instead time the Thread.delay call, with each thread accumulating the total time spent in Thread.delay, rather than the number of calls made to Thread.delay. The test passes if the total time spent by each thread in Thread.delay is within ±1.5 intervals of the overall test runtime, which is less sensitive to the resolution of Thread.delay.

I hit this while doing some CI cleaning for 4.x, where adding macOS ARM testing to CI instantly hits this problem with 4.14. I'd therefore like either to "fix" the test (as proposed here) or permanently remove it from 4.x and 5.x.

tmcgilchrist commented:
Looks like a reasonable fix to me. 

When I ran the entire test suite as part of the review I got a failure in `tests/lib-threads/delayintr.ml` which seems unrelated and hard to reliably reproduce.

``` shell
Running tests from 'tests/lib-threads' ...
 ... testing 'backtrace_threads.ml' => passed
 ... testing 'bank.ml' => passed
 ... testing 'beat.ml' => passed
 ... testing 'bufchan.ml' => passed
 ... testing 'close.ml' => passed
 ... testing 'delayintr.ml' => failed
 ... testing 'delayintr.ml' with line 4 (hassysthreads) => passed
 ... testing 'delayintr.ml' with line 5 (libunix) => passed
 ... testing 'delayintr.ml' with line 8 (setup-ocamlc.byte-build-env) => passed
 ... testing 'delayintr.ml' with line 11 (ocamlc.byte) => passed
 ... testing 'delayintr.ml' with line 14 (ocamlc.byte) => passed
 ... testing 'delayintr.ml' with line 15 (check-ocamlc.byte-output) => passed
 ... testing 'delayintr.ml' with line 16 (run) => passed
 ... testing 'delayintr.ml' with line 17 (check-program-output) => failed (program output /Users/tsmc/code/ocaml/ocaml-trunk/testsuite/tests/lib-threads/_ocamltest/tests/lib-threads/delayintr/ocamlc.byte/delayintr.byte.output differs from reference /Users/tsmc/code/ocaml/ocaml-trunk/testsuite/tests/lib-threads/delayintr.reference: 
--- /Users/tsmc/code/ocaml/ocaml-trunk/testsuite/tests/lib-threads/delayintr.reference	2024-03-13 13:37:07
+++ /Users/tsmc/code/ocaml/ocaml-trunk/testsuite/tests/lib-threads/_ocamltest/tests/lib-threads/delayintr/ocamlc.byte/delayintr.byte.output	2024-07-30 12:28:41
@@ -1 +1 @@
-Received signal early
+Received signal late

)
 ... testing 'delayintr.ml' with line 20 (setup-ocamlopt.byte-build-env) => passed
 ... testing 'delayintr.ml' with line 23 (ocamlopt.byte) => passed
 ... testing 'delayintr.ml' with line 26 (ocamlopt.byte) => passed
 ... testing 'delayintr.ml' with line 27 (check-ocamlopt.byte-output) => passed
 ... testing 'delayintr.ml' with line 28 (run) => passed
 ... testing 'delayintr.ml' with line 29 (check-program-output) => passed
```

dra27 committed:
Fix tests/lib-threads/beat.ml on macOS ARM

Posix specifically notes that nanosleep may delay the calling thread for
_longer_ than the requested timeout while also stating that it may only
return early on error or owing to delivery of a signal.

macOS (especially on Apple Silicon) appears to take a very liberal view
of sub-second nano-sleep calls, such as the the 0.33333 and 0.5 delays
in the beat.ml test.

The test is tweaked to time each Thread.delay (using Unix.gettimeofday)
and allow any extra delay to be credited against the next delay. An
unfair OS scheduler (and a malfunctionaling OCaml runtime) thus will
both still trigger a failure, but a slightly lazy OS scheduler which
inflates some of the nanosleep calls by 30% or so will not cause the
test to fail.

dra27 commented:
I don't know - that test also has a `Thread.delay` in it! Although at a cursory glance, it'd be surprising if it's actually the problem.

I liked this approach last night, then having slept on it I realise that by having the wallclock check around `Thread.delay` (in OCaml) rather than surrounding the `nanosleep` (in C) that we're not just checking `nanosleep` but also the reacquisition of the runtime lock. While my car was failing its annual MOT, I went down an incorrect rabbit-hole putting the timing code directly in C instead (https://github.com/dra27/ocaml/commit/6e85f3c5c2ba73eec646d399826e86beaf0907af). However, it becomes a very big test and it's not even technically testing `Thread.delay` anymore. While I was walking home cursing the state of the UK's roads and the damage they've done to my car, it occurred to me that there was a simpler way, which I've pushed.

The first approach - recording the _total_ time spent in `Thread.delay` (including or not including the time spent context switching _in the OCaml runtime_) instead of the number of iterations (as before) allows a parasitic scheduling case to pass the test, where before it didn't. Taking the 0.33333 thread, assume that its first call to `nanosleep` takes 0.4s, but that for some reason either the OCaml runtime (or the OS) then starves that thread for another 2.5s, pausing it just after its call the `Thread.delay` has returned but _before_ `Unix.gettimeofday` has measured the wallclock time. Then just before the end of the test (at 3s) the OS (or OCaml) allows that thread to run again, so it merrily records that its _single_ 0.33333 `Thread.delay` call took 2.9s and the test passes. Oops.

The new approach still uses counters, so it's still expecting one thread to increment that counter 9±1 times and the other to increment its counter 6±1 times. The difference is that each iteration is now permitted to knock-off up to 95% of the delay to account for the _previous_ `Thread.delay` having taken longer (i.e. if one `Thread.delay 0.5` takes 0.6s, the next delay will only be 0.4s to try to make up for it). Much simpler - no C required - and now the test will still catch either the OS or, more importantly, OCaml doing something strange with thread scheduling.

gasche commented on testsuite/tests/lib-threads/beat.ml:
```diff
@@ -18,10 +13,30 @@
+  let rec loop delta =
+    let start = Unix.gettimeofday () in
+    Thread.delay (delay -. delta);
+    let wallclock_delay = (Unix.gettimeofday () -. start) +. delta in
```
It is not obvious why you add `delta` back here, what this is supposed to mean.

dra27 replied:
My variable name's ended up unclear, I think. It's trying to account for how much over the requested slice we _still_ are. Slightly easier with an example: suppose the first call (so `delta = 0.0`) to `Thread.delay` for `delay = 0.5` in fact sleeps for 0.6 seconds. On the next iteration, `delta = 0.1`, so the thread requests `Thread.delay 0.4` which in fact sleeps for 0.5 seconds - wallclock_delay is then the 0.5 seconds we actually measured _plus_ the 0.1 seconds extra from the previous iteration (and the third iteration will try `delta = 0.2`).

gasche reviewed (commented):
I tried reviewing this to help but I found myself lukewarm about the result.

## Web search

I looked online for other bug reports about nanosleep on osx. I found these:
- https://stackoverflow.com/questions/48285535/macos-nanosleep-sleeping-for-far-too-long
- https://travis-ci.community/t/sleep-functions-are-not-accurate-on-macos/6122
- https://github.com/travis-ci/osx-image-bootstrap/blob/ccf3f7bb0fd34592b32e58b4d3273c71ba6ae3e1/bootstrap.sh#L32

These pages suggest that there is some way to ask OSX to stop overshooting wildly on `sleep` calls. (My guess: OSX believes that most applications that call `sleep` in a loop are badly written and makes an conscious effort to give them less resources. But there is an explicit opt-out for people who really need reliable sleep durations.)

## Deep down

The test is testing that `Thread.delay` works as expected, and it is currently disabled (on all systems?) because `Thread.delay` does *not* work as expected on OSX.

Your proposed fix is to write code in a sensibly more complex way to work around the issues with `Thread.delay` on OSX. But this does not seem to be solving the actual issue, which is that `Thread.delay` does not work as expected on OSX. If we want to re-enable the test and declare success, I think that we should:
1. Find a way to have `Thread.delay` work correctly on OSX, or
2. Document that `Thread.delay` does not work correctly, and possibly give recommendations on how to use it more reliably.

Maybe your last iteration is a good approach to do (2) (on first read I don't understand it well enough to tell), but then I think that instead of being hidden in an obscure test we should document it in thread.mli and let users know about it. Or maybe we don't want to do this, because it is too complex/arcane/fragile, but then that suggests that "fixing" the test in this way is not really a fix.

Does that make sense?

My impression from afar is that, barring a magical way to make `sleep` work again, we should disable the test *on OSX*, document the flaws of `Thread.delay` and `Unix.sleepf` somewhere, and move on.

dra27 commented:
It does make sense - I'd also found those various links (although also suggesting that on Travis in particular I think the settings didn't help). To be honest, after the C rabbit-hole I was veering towards deleting the test!

The test in 5.x is indeed completely disabled because of macOS (although I think there may have been some Risc-V instances of this test failing as well). Part of the reason I've chosen to try to tackle this is that it was previously thought to be a multicore-introduced problem, but it's very clearly always been the case - whichever outcome we end up with, I'd like to back-port it to 4.14's testsuite so that macOS arm64 CI can be switched on there too (at least while we're still maintaining it).

I'm not sure that `Thread.delay` is strictly misbehaving - I'd be quite happy with the idea that we clarify that `Thread.delay` delays for _at least_ the specified timeout.

Another option would of course be to restrict the test to non-macOS ... as a smoke test for something being seriously wrong in the runtime (i.e. threads being randomly starved of execution time) it seems possibly useful?

I'm now even less keen on my own fix having just seen a test run with it on 4.14 fail on amd64 macOS giving n1 = 11, n2 = 6 - i.e. that looks like that 3s delay in the main thread was much longer than expected!

ghost commented:
My EUR .02 on this:
- it does not make sense for OCaml to try to be smarter than `nanosleep` or any other sleep function. All sleep functions are documented to sleep for _at least_ the requested duration (unless signals are delivered), and it is not OCaml's job to try and do a more accurate job because OCaml is not an operating system. Therefore, changing the test to allow for fewer `Thread.delay` calls as long as the test duration is correct, is the right thing to do.
- `Unix.gettimeofday` is unfortunately not reliable as a time reference, because the clock may be changed (either manually or with `rdate` or `ntp`) while the test runs (even though it is extremely unlikely). It would be useful to make `caml_time_counter` available to OCaml code so that the test could use it as a reliable monotonic time counter for such purposes.

OlivierNicole commented:
Considering that this seems to be stalled, maybe the best solution is the not-completely-satisfactory one which is an increment over the current situation?

NickBarnes commented:
I agree with @gasche that `thread.mli` should be updated to reflect the actual semantics of `Thread.delay`: we can't guarantee on any mainstream platform that a sleeping thread will wake up in any given amount of time (for instance, if the machine is heavily loaded). OSX seems to be an easy demonstration of this. Then this test, as updated, can test something interesting about the actual semantics of `Thread.delay`.

xavierleroy commented:
I agree this test is cursed and should probably remain deactivated.

Historically, it was written for the "green thread", bytecode-level thread library, as a test for the scheduling of `Thread.delay` operation.  It was an interactive tests: I would run it and check visually that the "a" and "b"'s were printed in an acceptable 3-for-2 rhythm.  Then we tried to automate the test and it never really worked...

These days we rely on the OS to schedule threads, so there's really nothing to check at the OCaml level.

dra27 committed:
Remove the test completely

dra27 commented:
I've proposed an update to the documentation in #14948. I was updating the test on my laptop this morning to be non-macOS and then when I ran it (on Linux) it failed. So I finally take the hint - this PR now permanently fixes the test with `rm -f`

dra27 commented:
Given that the PR simply deletes a test, I don't think it necessary to wait for the two lagging CI runners.
